common/transport/amqp/src/amqp.ts

```diff
@@ -135,6 +88,280 @@ export class Amqp {
+                done(err);
+              } else {
+                this._senders[endpoint].removeListener('error', attachErrorHandler);
+                this._senders[endpoint].on('error', (err) => {
+                  debug('error on sender link: ' + endpoint + ': ' + err.toString());
+                  delete(this._senders[endpoint]);
+                });
+                debug('sender link attached: ' + endpoint);
+                done(null, this._senders[endpoint]);
+              }
+            });
+          },
+          detachReceiverLink: (endpoint: string, detachCallback: GenericAmqpBaseCallback<any>): void => {
+            if (!this._receivers[endpoint]) {
+              this._safeCallback(detachCallback);
+            } else {
+              this._detachLink(this._receivers[endpoint], (err?) => {
+                delete(this._receivers[endpoint]);
+                this._safeCallback(detachCallback, err);
+              });
+            }
+          },
+          detachSenderLink: (endpoint: string, detachCallback: GenericAmqpBaseCallback<any>): void => {
+            this._detachLink(this._senders[endpoint], (err?) => {
+              delete(this._senders[endpoint]);
+              this._safeCallback(detachCallback, err);
+            });
+          }
+        },
+        disconnecting: {
+          _onEnter: (disconnectCallback, err) => {
+            const disconnectFunc = err ? (callback) => {
+              callback();
+            } : (callback) => {
+              this._amqp.removeAllListeners('disconnected');
+              /*Codes_SRS_NODE_COMMON_AMQP_16_004: [The disconnect method shall call the done callback when the application/service has been successfully disconnected from the service] */
+              this._amqp.disconnect().then(() => {
+                callback();
+                return null;
+              }).catch((err) => {
+                callback(err);
+              });
+            };
+
+            const detachFunc = err ? (link, callback) => {
+              if (link) {
+                link.forceDetach(err);
+              }
+              callback();
+            } : (link, callback) => {
+              if (link) {
+                link.detach(callback);
+              } else {
+                callback();
+              }
+            };
+
+            detachFunc(this._cbs, () => {
+              let remainingLinks = this._getRemainingLinks();
+              /*Codes_SRS_NODE_COMMON_AMQP_16_034: [The `disconnect` method shall detach all open links before disconnecting the underlying AMQP client.]*/
+              async.each(remainingLinks, detachFunc, () => {
+                disconnectFunc((err) => {
+                  if (err) {
+                    this._fsm.transition('disconnected', disconnectCallback, err);
+                  } else {
+                    this._fsm.transition('disconnected', disconnectCallback);
+                  }
```

Comment:
style question:  This block operates exactly the same as if you just called transition with 3 parameters without checking the value of err.  Is it better to be explicit or terse? (I don't know the answer)

Reply:
I'm not sure I understand what you mean

---
In reply to: [139532979](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139532979) [](ancestors = 139532979)

Reply:
It's a silly thing, but as written, if err is defined, it calls 
  this._fsm.transition('disconnected', disconnectCallback, err)
If err is falsy, it calls 
  this._fsm.transition('disconnected', disconnectCallback).

You could get the same behavior by calling 
  this._fsm.transition('disconnected', disconnectCallback, err)

Because, if err is falsy, 
  this._fsm.transition('disconnected', disconnectCallback, err)
is the same as 
  this._fsm.transition('disconnected', disconnectCallback)

(but only if the transition function treats all falsy values the same, which isn't a horrible assumption.)

It's OK to ignore this comment.  I won't be offended.


---
In reply to: [139561201](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139561201) [](ancestors = 139561201,139532979)

Reply:
got it.

---
In reply to: [139807575](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139807575) [](ancestors = 139807575,139561201,139532979)